src/phases/reading.rs

```diff
@@ -0,0 +1,345 @@
+use std::{
+    collections::HashMap,
+    io::{self, Cursor, Write},
+};
+
+use citro2d_sys::{C2D_AlignCenter, C2D_AlignLeft, C2D_Text};
+use ctru::prelude::{Apt, Gfx, Hid, KeyPad};
+use libdoodle::{
+    blocks::{common1, miistd1::MiiData},
+    bpk1::{BPK1Blocks, BPK1File},
+    files::letter::Letter,
+};
+
+use crate::{
+    control_flow::{AbortMigration, MigrationFlow},
+    extdata::{ExtdataArchive, SwapdoodleRegion},
+    friend_list::{self, MiiMap},
+    gui::{
+        Gui, TOP_SCREEN_WIDTH,
+        scrollable_view::{ScrollableView, ScrollableViewData},
+    },
+    phases::process,
+    read::ReadExt,
+};
+
+pub struct ReadResult {
+    pub friends: MiiMap,
+    pub doodles: MiiMap,
+}
+
+pub fn reading<'a>(apt: &'a Apt, gfx: &'a Gfx, hid: &'a mut Hid, gui: &'a mut Gui) -> Scene<'a> {
+    Scene::new(apt, gfx, hid, gui)
+}
+
+fn friendly_read_data(extdata: &ExtdataArchive) -> (MiiMap, MiiMap) {
+    print!("Reading your friend list... ");
+    _ = io::stdout().flush();
+
+    let friends = friend_list::load_friend_list();
+    println!("done!");
+
+    println!("Reading your Swapdoodle extdata... ");
+
+    println!("Reading file /letter/manage.bin...");
+    let mut manage = BPK1Blocks::new_from_bpk1_bytes(&extdata.read_manage()).unwrap();
+    let cominf = manage
+        .iter_mut()
+        .find(|k| k.name == c"COMINF0")
+        .expect("File /letter/manage.bin should have a COMINF0, but it doesn't!");
+
+    let mut cursor = Cursor::new(&cominf.data);
+
+    let mut unknown_pids = HashMap::<u32, u32>::new();
+
+    let count = cursor.read_u32_le().unwrap();
+    cursor.set_position(0x40);
+
+    for _ in 0..count {
+        let pos = cursor.position();
+        let common =
+            common1::CommonInfo::from_bytes(&(cursor.read_const_num_of_bytes::<0x40>().unwrap()))
+                .unwrap();
+        let sender_pid = common.sender_pid;
+
+        if !friends.contains_key(&sender_pid) {
+            let letter_key = cursor.read_u32_le().unwrap();
+            unknown_pids.insert(sender_pid, letter_key);
+        }
+
+        cursor.set_position(pos + 0x80);
+    }
+    println!("done.");
+
+    let mut doodles = HashMap::<u32, MiiData>::new();
+
+    unknown_pids.iter().for_each(|row| {
+        let key = *row.1;
+        if let Some(mii) = Letter::new_from_bpk1_bytes(&extdata.read_letter_index(key))
+            .unwrap()
+            .sender_mii
+        {
+            doodles.insert(*row.0, mii);
+        }
+    });
```

Comment:
Since we're dropping `unknown_pids` after this anyway, might as well consume it in this operation and take ownership of its contents by using `.into_iter()`:
```suggestion
    unknown_pids.into_iter().for_each(|(pid, key)| {
        if let Some(mii) = Letter::new_from_bpk1_bytes(&extdata.read_letter_index(key))
            .unwrap()
            .sender_mii
        {
            doodles.insert(pid, mii);
        }
    });
```